### DISCUSSION TITLE
Not able to set multiple cookies with middleware function

### DISCUSSION THREAD

**Discussion by Thamasky** [vercel/next.js#33782] [Help] [answered]
Hi all,

I'm having issues trying to set multiple cookies in the _middleware function. It works perfect on localhost, but when deployed on vercel it only sets one cookie. It could be due the fact that Vercel combines everything into one single set-cookie header.

Code that sets the cookies:
```
   res.cookie(CookieNames.VendorCode, vendorCode, {
      maxAge: 365 * 24 * 60 * 60,
      secure: process.env.NEXT_PUBLIC_VERCEL_ENV !== 'development',
      sameSite: 'none',
    });

    res.cookie(CookieNames.CountryCode, countryCode, {
      maxAge: 365 * 24 * 60 * 60,
      secure: process.env.NEXT_PUBLIC_VERCEL_ENV !== 'development',
      sameSite: 'none',
    });

```

I tried changing this behaviour by changing the `res.cookie()` statements to serializing the cookies first, and then appending them all to the headers like this:

```
   const cookies: string[] = [];

    cookies.push(
      serialize(CookieNames.VendorCode, vendorCode, {
        maxAge: 365 * 24 * 60 * 60,
        secure: process.env.NEXT_PUBLIC_VERCEL_ENV !== 'development',
        sameSite: 'none',
      }),
    );

    cookies.push(
      serialize(CookieNames.CountryCode, countryCode, {
        maxAge: 365 * 24 * 60 * 60,
        secure: process.env.NEXT_PUBLIC_VERCEL_ENV !== 'development',
        sameSite: 'none',
      }),
    );

    cookies.forEach((cookie) => {
      res.headers.append('Set-Cookie', cookie);
    });
```

This sets multiple set-cookie headers on localhost, but when deployed to Vercel it gets reduced to a single set-cookie header again with all cookies in it. However it only sets one cookie, so the problem persists.

Example set-cookie header on Vercel:

```
set-cookie: ec-vc=ec; Max-Age=31536000; Secure; SameSite=None, ec-cc=be; Max-Age=31536000; Secure; SameSite=None
```

I also already tried to add domain & path to the cookies. But no difference there.

Any help on this would be really appreciated.

Kind regards,
Thomas

**Comment by ardaerzin** [comment]
we noticed a problem with a middleware related functionality last week, which was related to this issue. it works without any problems on local, but once deployed to vercel it stops functioning. here is a [reproduction](https://github.com/ardaerzin/next-mw-cookie)

just go to `/?token=test` and inspect cookies and you'll see the issue

**Comment by MerlinMason** [comment]
Ahh yeah, got the same issue trying to set a `token` and `refresh_token`... Looking at the types for `NextResponse` it looks like setting more than one cookie isn't possible currently :(

**Comment by Thamasky** [comment, marked as answer]
Hi, as an alternative workaround, you could try [this answer](https://github.com/vercel/community/discussions/259#discussioncomment-2090179) 

I already reworked my code to have one json cookie so I didn’t test.

**Comment by abhisharkjangir** [comment]
Alternate solution could be:

`NextJS version: 13.2.4`


```javascript
import { NextResponse } from 'next/server';
export async function middleware(request) {
  const response = NextResponse.next();
  const cookies = [
    'utm_campaign=ZO_AQ; Max-Age=3600; Path=/; Domain=.xyz.com;',
    'utm_medium=cpc; Max-Age=3600; Path=/; Domain=.xyz.com;',
  ];
  // Join cookies using comma(,)
  response.headers.set('Set-Cookie', cookies.join(','));
  return response;
}
```
